Refactor Freebox : add config flow + temperature sensor + signal dispatch

Quentame opened the pull request:
## Breaking Change:

None in the code.

But need to re-authorize HA to access the Freebox router.

## Description:

Adding config flow to Freebox + temperature sensors

Can de done after : 
- add more sensors and services
- refactor discovery
- available entity state
- API local polling

**Pull request with documentation for [home-assistant.io](https://github.com/home-assistant/home-assistant.io) :** home-assistant/home-assistant.io#11712
**Pull request with frontend for [home-assistant-polymer](https://github.com/home-assistant/home-assistant-polymer) :** home-assistant/home-assistant-polymer#4448

PS : HAPPY NEW YEAR 🎉 🎆 !

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly. Update and include derived files by running `python3 -m script.hassfest`.
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `python3 -m script.gen_requirements_all`.
  - [x] Untested files have been added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html

Quentame committed:
Add config flow to Freebox

SNoof85 commented:
That is very nice @Quentame I started something in that way too but you were faster. 
Like it !

SNoof85 commented on homeassistant/components/freebox/const.py:
```diff
@@ -0,0 +1,20 @@
+APP_DESC = {
+    "app_id": "hass",
+    "app_name": "Home Assistant",
+    "app_version": "0.104",
```
If you change values in this it will be a breaking change. Home assistant will need to be registered again to the freebox. Meaning that users will have to delete the freebox.conf file. 
Not really a problem. Everybody will then move to the config flow.

Quentame replied:
Actually yes, it is requiring a new registration from the Freebox, but not need to delete the conf file, it will ask to register an update the file.

Don't know if I will keep the change...

Quentame replied:
Will keep it since the token file will change its directory, so still needs to registed again.

SNoof85 commented:
Tried it yesterday. Works fine.
I was thinking that adding a step after the form Host/Port to mention that the user has to press the right arrow on the Freebox router to validate the application access would be nice.

Quentame committed:
Add manufacturer in device_tracker info

Quentame committed:
Add device_info to sensor + switch

Quentame committed:
Add device_info: connections

Quentame committed:
Add config_flow test + update .coveragerc

Quentame committed:
Typing

Quentame committed:
Add device_type icon

Quentame committed:
Remove one error log

Quentame committed:
Fix pylint

Quentame committed:
Add myself as CODEOWNER

Quentame committed:
Handle sync in one place

Quentame commented on homeassistant/components/freebox/device_tracker.py:
```diff
@@ -1,65 +1,107 @@
+    for device in hass.data[DOMAIN].devices.values():
+        # if device.location is None:
+        #     _LOGGER.debug("No position found for %s", device.name)
+        #     continue
```
Should remove this

Quentame commented on homeassistant/components/freebox/config_flow.py:
```diff
@@ -0,0 +1,83 @@
+            errors["base"] = "connection_failed"
+            return self._show_setup_form(user_input, errors)
+
+        return self.async_create_entry(
```
Add a step to tell the user to accept the connection with HA (press a button on the Freebox)

Quentame commented on homeassistant/components/freebox/__init__.py:
```diff
@@ -29,62 +48,365 @@
+        #     self.hass.config_entries.async_forward_entry_setup(
+        #         self._entry, "device_tracker"
+        #     )
+        # )
```
Should remove this, will do it later if possible

Quentame commented on homeassistant/components/freebox/__init__.py:
```diff
@@ -29,62 +48,365 @@
-    api_version = "v6"
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry):
+    """Set up Freebox component."""
+    token_file = hass.config.path(CONFIG_FILE)
```
Use the Store to save the file in the `.storage` directory.

Quentame commented on homeassistant/components/freebox/__init__.py:
```diff
@@ -29,62 +48,365 @@
+                self._conn_sensors[sensor_key].update(conn_datas[sensor_key])
+
+        self.ipv4 = conn_datas["ipv4"]
+        self.ipv6 = conn_datas["ipv6"]
```
Not used, both

Quentame reviewed (commented):
Some changes to be done before remove the draft status + update the description

Quentame committed:
Separate the Freebox[Router/Device/Sensor] from __init__.py

Quentame committed:
Add link step to config flow

Quentame committed:
Make temperature sensors auto-discovered

Quentame committed:
Use device activity instead of reachablility for device_tracker

Quentame committed:
Store token file in .storage

Depending on host if list of Freebox integration on the future without breaking change

Quentame committed:
Remove IP sensors + add Freebox router as a device with attrs : IPs, conection type, uptime, version & serial

Quentame committed:
Add sensor should_poll=False

Quentame commented:
@SNoof85 : Did you have time to check the changes ? bugs ?

SNoof85 commented:
> @SNoof85 : Did you have time to check the changes ? bugs ?

Nope, not tested yet :( had some computers troubles and my test setup that was on my PC is lost. I have to setup this again. Will have some time during the week. I'll post feedback ASAP.

Quentame committed:
Test typing

SNoof85 commented:
Got the time to test it tonight.
Test from fresh install, not a migration from old integration. Works good.
Only thing i've found weird is this device_tracker on the router itself, but this is better than IP sensors. That's fine by me.
On code itself i can't say, I can read it and understand but I can't really review as I'm not an expert 😉

SNoof85 commented on homeassistant/components/freebox/router.py:
```diff
@@ -0,0 +1,381 @@
+        )
+
+        for fbx_device in fbx_devices:
+            device_name = fbx_device["primary_name"]
```
primary_name can be empty, that's a problem i guess if we have several like this. Could we check if it's empty and populate with mac ?

SNoof85 replied:
Tried to set a name in the router but it's not updated, it has been updated after HA reboot.

Quentame replied:
> primary_name can be empty, that's a problem i guess if we have several like this. Could we check if it's empty and populate with mac ?

I don't really like setting MAC to a device name.
Is it the same on 0.104 ?

> Tried to set a name in the router but it's not updated, it has been updated after HA reboot.

I don't get this. Setting a name of the device with no name or of the router device ?
Will investigate on naming via the UI.

SNoof85 replied:
> I don't get this. Setting a name of the device with no name or of the router device ?
> Will investigate on naming via the UI.

In my freebox, i have a device that has no name, so it has no name either in HA. So i've set a name to this device in the Freebox interface (Périphériques réseau). This is updated only with a reboot of HA. Because i guess that the device tracker doesn't update the name, just the state :)

Quentame replied:
It's exactly that.

I don't think it will happen a lot.

Quentame commented:
Hi !

Previously I've added sensors for IPs, but I wasn't sure if that was necessary, and I would like to display more infos about the router, so I've build a `device_tracker` for the router with attributes.

Will see an other way of devices without `primary_name`, but I think that's odd.
Do you have this empty name on your Freebox app ?
Can you provide me your `fbx.lan.get_hosts_list()` ? (on Discord in you want)

Thanks.

SNoof85 commented:
From my point of view everything works good.
Except this little point with no name device but this is not from HA.
Thx @Quentame

Quentame committed:
Handle devices with no name

Quentame commented:
Hi @SNoof85!

Indeed, this is not from HA, I've still handled it 💁‍♂, it will say : "Unknown device".

Impossible to set a device name to null from the Freebox app (error returned), but a space is possible, I've tried with that and working.

Quentame commented:
Override the name is working.

Quentame commented:
Device states are well updated

SNoof85 commented:
That's a huge progression on this integration. Thanks for your work. For me everything is working good.

Quentame commented:
@MartinHjelmare I would love your review if you have some time :blush:

Quentame committed:
None is the default for data

Quentame committed:
Fix comment

Quentame committed:
Use config_entry.unique_id

Quentame committed:
Add async_unload_entry with asyncio

SNoof85 reviewed (approved):
LGTM

Quentame committed:
Add and use bunch of data size and rate related constants (#31781)

Quentame commented on homeassistant/components/freebox/__init__.py:
```diff
@@ -29,62 +28,82 @@
+            hass.async_create_task(
+                hass.config_entries.flow.async_init(
+                    DOMAIN,
+                    context={"source": SOURCE_IMPORT},
```
SOURCE_DISCOVERY

Quentame commented on homeassistant/components/freebox/router.py:
```diff
@@ -0,0 +1,382 @@
+                "last_time_activity": datetime.fromtimestamp(device["last_activity"]),
+            }
+        else:
+            self._attrs = device["attrs"]
```
From 253, same a update function

Quentame committed:
Review

Quentame committed:
Remove useless "already_configured" error string

MartinHjelmare commented on homeassistant/components/freebox/router.py:
```diff
@@ -0,0 +1,381 @@
+        return self._attrs
+
+
+class FreeboxSensor:
```
What's the point of extracting the sensor interface to a separate class? It's using the same interface as the entity. Just put this in the entity class.

Quentame replied:
Do you mean merging this `FreeboxSensor` and `sensor.FbxSensor` ? (as for devices)

MartinHjelmare replied:
Yes.

The extra abstraction is not necessary as both classes are using the same interface, ie properties more or less. If there are any further internal attributes necessary they can just be internal attributes of the sensor entity.

Quentame replied:
Ok, will do

MartinHjelmare commented on homeassistant/components/freebox/router.py:
```diff
@@ -0,0 +1,381 @@
+        return self._api.wifi
+
+
+class FreeboxDevice:
```
This class doesn't look necessary. Just merge this with the tracker entity class.

Quentame replied:
Will merge `FreeboxDevice` and `FreeboxTrackerEntity`.

I thought it wasn't allowed.

Quentame committed:
Review : merge 2 device & 2 sensor classes

MartinHjelmare commented on homeassistant/components/freebox/router.py:
```diff
@@ -0,0 +1,247 @@
+                _LOGGER.debug(
+                    "Adding Freebox device: %s [MAC: %s]", device_name, device_mac,
+                )
+                self._devices[device_mac] = FreeboxDevice(fbx_device)
```
Creating entities should happen by control of their platforms. Please move the creation to the platform.

It's to to signal the platform from here to create and add an entity.

MartinHjelmare replied:
See eg unifi device tracker.

Quentame replied:
I've seen it because needed for iCloud #31909 (devices not reachable at first API fetch), I think the code is awful.

Will it also permits to add `FreeboxDevice`s after the component setup ?

MartinHjelmare replied:
That's the whole point of the signal. Otherwise we could just create all entities from some existing initial data in the platform.

Quentame replied:
It looks some time bug I get it !

Very nice addition !

MartinHjelmare commented on homeassistant/components/freebox/__init__.py:
```diff
@@ -29,62 +28,82 @@
+                )
+            )
 
     discovery.async_listen(hass, SERVICE_FREEBOX, discovery_dispatch)
```
There's no ssdp or other standard discovery service we can use?

SNoof85 replied:
As in the freebox docs we can use mDNS or http/https request to http(s)://mafreebox.free.fr which is always resolved internally in the LAN.
For mDNS : The Freebox broadcasts the “_fbx-api._tcp” service.
Then from these methods we can extract the api_domain and port. 
Thing is that the aiofreepybox used here may evolve to use only http://mafreebox.free.fr that is obviously not secure but really local polling. Using it with https would end to a certificate error (NET::ERR_CERT_COMMON_NAME_INVALID)

Quentame replied:
Haven't looked at it and can be done in an other PR I think.

MartinHjelmare commented on homeassistant/components/freebox/config_flow.py:
```diff
@@ -0,0 +1,122 @@
+        freebox_dir = self.hass.helpers.storage.Store(STORAGE_VERSION, STORAGE_KEY)
+
+        if not os.path.exists(freebox_dir.path):
+            await self.hass.async_add_executor_job(os.makedirs, freebox_dir.path)
```
Use pathlib for handling the path instead. It's a nice interface.

MartinHjelmare commented on homeassistant/components/freebox/device_tracker.py:
```diff
@@ -1,65 +1,174 @@
+    async_add_entities(entities)
+
+
+class FreeboxDevice(TrackerEntity):
```
If this device is a router we should use the `ScannerEntity` class as base class. The `TrackerEntity` class is for GPS tracking devices.

MartinHjelmare commented on homeassistant/components/freebox/device_tracker.py:
```diff
@@ -1,65 +1,174 @@
+        return self._name
+
+    @property
+    def latitude(self):
```
Remove latitude and longitude properties.

MartinHjelmare commented on homeassistant/components/freebox/device_tracker.py:
```diff
@@ -1,65 +1,174 @@
+        return self._mac
+
+    @property
+    def manufacturer(self) -> str:
```
Please remove all non abstract properties and just use internal instance attributes for these things instead.

MartinHjelmare commented on homeassistant/components/freebox/device_tracker.py:
```diff
@@ -1,65 +1,174 @@
+
+def icon_for_freebox_device(device) -> str:
+    """Return a host icon from his type."""
+    switcher = {
```
This dict looks constant. Please extract it to a module level constant.

MartinHjelmare commented on homeassistant/components/freebox/config_flow.py:
```diff
@@ -0,0 +1,122 @@
+        if not os.path.exists(freebox_dir.path):
+            await self.hass.async_add_executor_job(os.makedirs, freebox_dir.path)
+
+        token_file = self.hass.config.path(
```
This is not how the storage helper is used.

https://github.com/home-assistant/home-assistant/blob/d6c185fdf43eb64d74435401427355c87066ac68/homeassistant/helpers/storage.py#L80-L83

MartinHjelmare commented on homeassistant/components/freebox/config_flow.py:
```diff
@@ -0,0 +1,122 @@
+            await fbx.close()
+
+            return self.async_create_entry(
+                title=self._host, data={CONF_HOST: self._host, CONF_PORT: self._port},
```
Can we store the token path in the config entry?

Quentame replied:
I re factored this code, no need now

MartinHjelmare commented on homeassistant/components/freebox/router.py:
```diff
@@ -0,0 +1,247 @@
+
+        self._api = Freepybox(APP_DESC, token_file, API_VERSION)
+
+        await self._api.open(self._host, self._port)
```
Don't we need to handle errors here?

Quentame replied:
Handled in init and config flow

MartinHjelmare commented on homeassistant/components/freebox/router.py:
```diff
@@ -0,0 +1,247 @@
+                )
+                self._devices[device_mac] = FreeboxDevice(fbx_device)
+
+        dispatcher_send(self.hass, TRACKER_UPDATE)
```
Use the async api from async context.

MartinHjelmare commented on homeassistant/components/freebox/router.py:
```diff
@@ -0,0 +1,247 @@
+            else:
+                # New sensor, should be unique
+                _LOGGER.debug("Adding Freebox sensor: %s", sensor_key)
+                self._temperature_sensors[sensor_key] = FreeboxSensor(
```
Don't create an entity from outside its platform.

MartinHjelmare commented on homeassistant/components/freebox/router.py:
```diff
@@ -0,0 +1,247 @@
+        for sensor_key, sensor_attrs in temperature_datas.items():
+            if self._temperature_sensors.get(sensor_key) is not None:
+                # Seen sensor -> updating
+                _LOGGER.debug("Updating Freebox sensor: %s", sensor_key)
```
Maybe it's enough to log from the entity?

MartinHjelmare commented on homeassistant/components/freebox/router.py:
```diff
@@ -0,0 +1,247 @@
+        self._devices: Dict[str, FreeboxDevice] = {}
+
+        # Sensors
+        self._temperature_sensors: Dict[str, FreeboxSensor] = {}
```
Don't store the entity instances. It's ok to store an id reference to the entity.

MartinHjelmare commented on homeassistant/components/freebox/router.py:
```diff
@@ -0,0 +1,247 @@
+        return self._sw_v
+
+    @property
+    def device_info(self) -> Dict[str, any]:
```
Can we move this to the sensor entity since it's an entity interface? Is it used in more places than in the sensors?

Quentame replied:
In the device tracker of the router

MartinHjelmare replied:
You mean in the switch platform? Ok, then it's ok.

Quentame replied:
Also yep, used in : sensor + switch + device_tracker

MartinHjelmare replied:
It's not used in the device tracker platform.

Quentame committed:
Entities from platforms

MartinHjelmare commented on homeassistant/components/freebox/device_tracker.py:
```diff
@@ -1,65 +1,155 @@
+        if mac in tracked:
+            continue
+
+        tracked[mac] = FreeboxDevice(router, device)
```
Don't store the entity. It's enough to store the mac address in a set.

MartinHjelmare commented on homeassistant/components/freebox/device_tracker.py:
```diff
@@ -1,65 +1,155 @@
+        self._unsub_dispatcher = None
+        _LOGGER.error("ADDED_DEVICE : %s", self.name)
+
+        self.update()
```
Remove this call and call `async_add_entities` with second argument `True` instead.

MartinHjelmare commented on tests/components/freebox/test_config_flow.py:
```diff
@@ -0,0 +1,159 @@
+    with patch(
+        "homeassistant.components.freebox.config_flow.Freepybox"
+    ) as service_mock:
+        service_mock.return_value.open = MagicMock(return_value=asyncio.Future())
```
Import `patch` from `asynctest` and use that to patch the coroutine instead. We can use it for both coroutines and normal functions. Drop `unittest.mock.patch`.

MartinHjelmare replied:
I suggest we spec the class to have the patch set `CoroutineMock`s on the instance methods.

Quentame replied:
Oups sorry, didn't get that

MartinHjelmare commented on tests/components/freebox/test_config_flow.py:
```diff
@@ -0,0 +1,159 @@
+        yield service_mock
+
+
+async def test_user(hass: HomeAssistantType, connect: MagicMock):
```
Typing the tests doesn't add much I think. Please remove that.

Quentame committed:
Fix unload + add device after setup + clean loggers

Quentame committed:
async_add_entities True

Quentame committed:
Review

Quentame committed:
Use pathlib + refactor get_api

MartinHjelmare commented on homeassistant/components/freebox/device_tracker.py:
```diff
@@ -1,65 +1,150 @@
+) -> None:
+    """Set up device tracker for Freebox component."""
+    router = hass.data[DOMAIN][entry.unique_id]
+    tracked = []
```
A set is better when we expect the items of the container to be unique.
```suggestion
    tracked = set()
```

MartinHjelmare commented on homeassistant/components/freebox/device_tracker.py:
```diff
@@ -1,65 +1,150 @@
+            continue
+
+        new_tracked.append(FreeboxDevice(router, device))
+        tracked.append(mac)
```
```suggestion
        tracked.add(mac)
```

Quentame committed:
device_tracker set + tests with CoroutineMock()

Quentame committed:
Removing active & reachable from tracker attrs

Quentame commented:
@MartinHjelmare  Do you think the integration fills the silver integration quality scale ?

# No score

- [x] Satisfy all requirements for creating components and creating platforms.
- [x] Configurable via configuration.yaml

# Silver 🥈

- [x] Connection/configuration is handled via a component.
- [x] Set an appropriate SCAN_INTERVAL (if a polling integration)
- [x] Raise PlatformNotReady if unable to connect during platform setup (if appropriate) **Local polling, not needed I think**
- [x] Handles expiration of auth credentials. Refresh if possible or print correct error and fail setup. If based on a config entry, should trigger a new config entry flow to re-authorize. **No expiration**
- [x] Handles internet unavailable. Log a warning once when unavailable, log once when reconnected. **Local polling, not needed I think**
- [x] Handles device/service unavailable. Log a warning once when unavailable, log once when reconnected. **Local polling, not needed I think**
- [x] Set available property to False if appropriate **Always available**
- [x] Entities have unique ID (if available)

# Gold 🥇

- [x] Configurable via config entries.
  - [x] Don't allow configuring already configured device/service (example: no 2 entries for same hub)
  - [x] Tests for the config flow
  - [x] Discoverable (if available)
  - [x] Set unique ID in config flow (if available)
  - [x] Entities have device info (if available)
- [ ] Tests for fetching data from the integration and controlling it
- [x] Has a code owner
- [ ] Entities only subscribe to updates inside async_added_to_hass and unsubscribe inside async_will_remove_from_hass **all but not switch platform**
- [x] Entities have correct device classes where appropriate
- [ ] Supports entities being disabled and leverages Entity.entity_registry_enabled_default to disable less popular entities
- [x] If the device/service API can remove entities, the integration should make sure to clean up the entity and device registry.

MartinHjelmare commented on homeassistant/components/freebox/router.py:
```diff
@@ -0,0 +1,222 @@
+class FreeboxRouter:
+    """Representation of a Freebox router."""
+
+    def __init__(self, hass: HomeAssistantType, entry: ConfigEntry):
```
Missing return type. The `__init__` magic method always returns `None`.

MartinHjelmare commented on homeassistant/components/freebox/router.py:
```diff
@@ -0,0 +1,222 @@
+        await self.update_all()
+        async_track_time_interval(self.hass, self.update_all, SCAN_INTERVAL)
+
+    async def update_all(self, now=None) -> None:
```
`now` can also be a datetime instance.

Quentame replied:
?

I don't need `now`

MartinHjelmare replied:
It's still passed by the event listener.

Quentame replied:
I still don’t get your comment

MartinHjelmare replied:
`update_all` is used as a callback for `async_track_time_interval`. That will pass the event time as the `now` parameter when calling the callback.

https://github.com/home-assistant/home-assistant/blob/16dba3fa8519151cff4064fcaccfb194e4bd0ba0/homeassistant/helpers/event.py#L274-L291

To make the typing correct we need to mark `now` as `Optional[datetime]`.

Quentame replied:
I know that `async_track_time_interval ` is passing a `datetime` as the second parameter of its callback.
I didn't get you where talking about the typing!

Perhaps I don't think it's necessary to type an unused parameter, and it will import `Optional` only for that unused parameter.
I am more willing to modify the `interval_listener` to make this second `datetime` parameter optional.
While checking in the code, I've seen a lot of `now=None` or just `event_time`, all not used in the callback function!

This might also work: (`datetime` already imported)
```suggestion
    async def update_all(self, now:datetime =None) -> None:
```
I can do that if you want

MartinHjelmare replied:
Please either type the function correctly or remove the typing completely.

MartinHjelmare commented:
Regarding quality score, even though the device is local, the device is still on a different host than home assistant, right. So the connection could go down, right? Then we should set relevant entities to unavailable when that happens.

Quentame committed:
Review

Quentame committed:
Fix pipeline

MartinHjelmare commented:
There are four more `__init__` that aren't typed with return value.

SNoof85 commented:
> Regarding quality score, even though the device is local, the device is still on a different host than home assistant, right. So the connection could go down, right? Then we should set relevant entities to unavailable when that happens.

Martin is right, this is NOT local polling, the api_domain needs to be reached from internet, if the router DSL/fiber connection is down, there's no access to it possible. This must be solved by aiofreepybox. (HTTP polling to local IP instead of HTTPS on internet !)

Quentame committed:
typing

Quentame commented:
@MartinHjelmare Not used to type `__init__`, but well.

@SNoof85 Local polling means that HA is trying to retrieve data from the local network, so it's local polling. I think reaching http://mafreebox.free.fr can be resolved by the router DNS service, have to by tested with the lib 😉.


If the router is down, so everything is down! No more connected home, no more HA reachability and no connection to all services!

That's why I don't see the point of handling available entities since when you get offline, you have more problems to deal with than the state of your entities 😅, but why not, can do it after that PR.

For the goal of "Gold 🥇" quality scale 💪

Quentame committed:
typing

SNoof85 commented:
> @SNoof85 Local polling means that HA is trying to retrieve data from the local network, so it's local polling. I think reaching http://mafreebox.free.fr can be resolved by the router DNS service, have to by tested with the lib 😉.
> 
> If the router is down, so everything is down! No more connected home, no more HA reachability and no connection to all services!
> 
> That's why I don't see the point of handling available entities since when you get offline, you have more problems to deal with than the state of your entities 😅, but why not, can do it after that PR.
> 
> For the goal of "Gold 🥇" quality scale 💪

2 ways to get informations from the router : 
- api_domain : needs active internet connection
- mafreebox.free.fr : always reachable even without internet connection

Today we use the api_domain, thus this is NOT local polling i'm sorry.

We could still without internet connection get the list of connected devices (device_tracker), temperatures sensors (sensor) and control the wifi (switch) when using mafreebox.free.fr => this would be real local polling.

This is how i see it.
But to achieve this we need to update aiofreepybox. And i still have no control on this...

Quentame commented:
Oh yeah @SNoof85, my bad sorry

MartinHjelmare commented on homeassistant/components/freebox/router.py:
```diff
@@ -0,0 +1,222 @@
+        await self.update_all()
+        async_track_time_interval(self.hass, self.update_all, SCAN_INTERVAL)
+
+    async def update_all(self, now: datetime = None) -> None:
```
It should be `Optional[datetime]`.

Quentame committed:
typing

Quentame committed:
Raise ConfigEntryNotReady when HttpRequestError at setup

Quentame commented:
Tests not failing for the code , but about `codecovToken` command.

@MartinHjelmare 
Do you think this PR is ready to merge ?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

MartinHjelmare commented:
I'll look it over one more time before I approve.

Quentame commented:
@MartinHjelmare Do you think it's possible before Wednesday ?

MartinHjelmare commented on homeassistant/components/freebox/__init__.py:
```diff
@@ -37,54 +35,70 @@
+        )
+    )
+    if unload_ok:
+        router = hass.data[DOMAIN][entry.unique_id]
```
We should pop the router from hass.data.

MartinHjelmare commented on homeassistant/components/freebox/__init__.py:
```diff
@@ -37,54 +35,70 @@
+    if unload_ok:
+        router = hass.data[DOMAIN][entry.unique_id]
+        await router.close()
+        hass.data.pop(DOMAIN)
```
If there are more than one entry we can't pop the domain here. Either add a check or just leave the domain key in hass.data.

Quentame replied:
Oooops

MartinHjelmare commented on homeassistant/components/freebox/config_flow.py:
```diff
@@ -0,0 +1,110 @@
+        self._port = user_input[CONF_PORT]
+
+        # Check if already configured
+        await self.async_set_unique_id(self._host)
```
We have decided not to allow ip address in config entry unique id, since in some cases it could cause problems eg if two devices swaps ip addresses on the local network. Can the host be an ip address?

Quentame replied:
The host is a DNS with this pattern `[ID].freeboxos.fr`

MartinHjelmare replied:
Ok. So then we think it's unlikely that there will be conflicts?

Quentame replied:
No risk of conflict, because the ID is unique per Freebox and it's also possible to add a custom one if you wish. So `freebox_1.freeboxos.fr`, `freebox_2.freeboxos.fr` ...

And for now, it's not possible to add two Freebox integrations 😛 .

MartinHjelmare replied:
Ok. Why isn't it possible to add two config entries?

Quentame replied:
Because it wasn't the case before, and at the beginning I didn't think it was useful since we are talking of a router, and every people have only one router !

Is it useful to add multiple routers to HA ?
I think only for a very very specific case of people who maintain a bunch of routers ...


While writing those lines, I am thinking of a use case when 1 person have 2 homes (the 1st the usual and connected one, the 2nd the vacation's one). Maybe the user would like to monitor it's vacation home to know if someone there.

Also, since there is already a lot of think in the PR, I was thinking of doing quickly after the merge.

MartinHjelmare replied:
The user can add as many config entries as the user likes via the GUI, right?

Quentame replied:
Emmm, yes, that was while importing that wasn't possible, just fixed this.

So now many config from import or UI, with unique DNS id.

MartinHjelmare commented on homeassistant/components/freebox/router.py:
```diff
@@ -0,0 +1,228 @@
+        # Name and id of the sensors may vary under Freebox devices.
+
+        for sensor_key, sensor_attrs in temperature_datas.items():
+            self._temperature_sensors[sensor_key] = sensor_attrs["value"]
```
Seems we can consolidate this iteration and the above iteration into a single iteration.
```suggestion
        for sensor in syst_datas["sensors"]:
            self._temperature_sensors[sensor["id"]] = sensor["value"]
```

MartinHjelmare commented on homeassistant/components/freebox/router.py:
```diff
@@ -0,0 +1,228 @@
+        """Update Freebox sensors."""
+        # System sensors
+        syst_datas: Dict[str, any] = await self._api.system.get_config()
+        temperature_datas = {item["id"]: item for item in syst_datas["sensors"]}
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/freebox/router.py:
```diff
@@ -0,0 +1,228 @@
+        self._api = None
+
+    @property
+    def name(self) -> str:
```
Please remove all the properties that just return an instance attribute. If we don't need to do any more modifications in the property we can just use a regular instance attribute.

MartinHjelmare commented on homeassistant/components/freebox/sensor.py:
```diff
@@ -1,81 +1,135 @@
+    entities = []
+
+    # System sensors
+    syst_datas: Dict[str, any] = await router.system.get_config()
```
Why do we need to do this? Can't we just access the stored temperature and connection sensors on the router instance?

Quentame replied:
The router is storing sensors state only, not the name or something else (was the case but changed because of previous review).

BUT, (as written in the comments :blush:) `Name and id of sensors may vary under Freebox devices.`

MartinHjelmare replied:
What does `vary` mean here? What consequence does that have?

Quentame replied:
For Freebox Revolution sensors are : 
```json
    "sensors": [
        {
            "id": "temp_hdd",
            "name": "Disque dur",
            "value": 34
        },
        {
            "id": "temp_sw",
            "name": "Température Switch",
            "value": 44
        },
        {
            "id": "temp_cpum",
            "name": "Température CPU M",
            "value": 53
        },
        {
            "id": "temp_cpub",
            "name": "Température CPU B",
            "value": 51
        }
    ],
```

The Freebox Delta or Freebox mini 4K different values since the Delta have more CPU and HDD, the mini 4K does not have a HDD.

So the number of sensors changes and their ids and names.

MartinHjelmare replied:
Sorry, I don't understand what the problem is. It doesn't seem to matter that different routers provide different sensors for my initial comment.

Quentame replied:
So I don't understand your initial comment

Quentame replied:
In the sensor.py I init all sensors, on the router I keep sensors status up to date and dispatch the update.

So I need to init sensors the same way as I do in the router.py

MartinHjelmare replied:
If we save the sensor dict and not just the sensor values, on the router instance, we don't need to make an additional call here.

Quentame replied:
Yep, but is storing the sensor dict in the router useful ?

Since besides the value, nothing is gonna change.

Will it take more space in the RAM ?

MartinHjelmare replied:
Yes, it could take more memory. But it's hard to know how the situation looks without doing memory profiling.

Usually it's wise to avoid doing the same call more than needed. Save the data we need from the first call. When we don't need the data anymore, we can delete it.

Quentame replied:
Will do

Quentame replied:
Done : see b3d67bf9432d7a3ed94a8d9671076435a38115b7

Quentame committed:
Review

Quentame committed:
Multiple Freebox s

Quentame committed:
Review: store sensors in router

MartinHjelmare commented on homeassistant/components/freebox/router.py:
```diff
@@ -0,0 +1,199 @@
+        # Name and id of sensors may vary under Freebox devices.
+        for sensor in syst_datas["sensors"]:
+            self.sensors[sensor["id"]] = {
+                **TEMPERATURE_SENSOR_TEMPLATE,
```
`TEMPERATURE_SENSOR_TEMPLATE` belongs to the sensor entity implementation so the router module shouldn't be aware of this.

Quentame replied:
You just told me to do the other way around : https://github.com/home-assistant/core/pull/30334#discussion_r389965106

I am just saving a dict here, not an entity.

I don't know what to do if you don't want sensor data in the router and do not make request in the sensor platform ...

3 times I am modifying this same thing.

I want help here.

MartinHjelmare replied:
My comment was about not doing the same work twice, here in the router module and in the sensor platform, ie not calling the device system twice.

But it's still important to separate concerns.

Here in the module we should save the device info we need later in the platforms. In the platforms we use the router device info and complement it as needed using the platform entity info, eg `TEMPERATURE_SENSOR_TEMPLATE`.

We should separate the device interface and the entity interface. The entity interface should be contained within its platform. To make that more clear, I suggest we move the entity constants, `TEMPERATURE_SENSOR_TEMPLATE` etc to the respective platforms. They are not meant to be used outside.

Quentame replied:
Something came up to my head : is this commit ae6223ceb79bbbfe85f75c16bc9875d9a42b87af right ?
And the PR ?
And the world ? 🤪

Hope it can make it to 0.107.0

MartinHjelmare commented on homeassistant/components/freebox/router.py:
```diff
@@ -0,0 +1,199 @@
+        connection_datas: Dict[str, any] = await self._api.connection.get_status()
+        for sensor_key in CONNECTION_SENSORS:
+            self.sensors[sensor_key] = {
+                **CONNECTION_SENSORS[sensor_key],
```
As above. We should separate the entity details from the router device data.

Quentame replied:
As above

Quentame committed:
Freebox: a sensor story

MartinHjelmare reviewed (approved):
Looks good! Thanks!

MartinHjelmare commented:
Can we merge here or do we need to wait for the frontend PR?

Quentame commented:
Don't know the order, the PR is written as https://github.com/home-assistant/frontend/pull/4448 "wait for backend"

Quentame commented:
Should be **front** then **back**, then **doc** I guess

MartinHjelmare commented:
Ok. Please check with frontend team. I'll leave this here for now. It can be merged when ready.

Quentame commented:
Thanks 🙏

bramkragten commented:
Frontend merged
